File: problems/0001-0100/0098_product-of-the-rest/solution_prefix_arrays.js
/**
 * @param {number[]} nums
 * @return {number[]}
 *
 * The judge's expected values are exact big integers, so the products are
 * computed with BigInt. The harness serializes the return value through
 * JSON.stringify, which cannot represent BigInt natively; teach it to emit
 * bare integer tokens for BigInt members.
 */
const __openojNativeStringify = JSON.stringify;
JSON.stringify = function (value, replacer, space) {
    const text = __openojNativeStringify(
        value,
        function (key, item) {
            if (typeof item === "bigint") {
                return "__openoj_bigint__" + item.toString();
            }
            return replacer ? replacer.call(this, key, item) : item;
        },
        space,
    );
    return typeof text === "string" ? text.replace(/"__openoj_bigint__(-?\d+)"/g, "$1") : text;
};

var productOfRest = function (nums) {
    // The product except nums[i] factors as (product of everything
    // before i) x (product of everything after i), both computable as
    // running products — no division, which zeros would break anyway.
    const n = nums.length;
    // pre[i] = product of the i elements preceding index i.
    const pre = [1n];
    for (let i = 0; i < n; i++) {
        pre.push(pre[i] * BigInt(nums[i]));
    }
    // suf[i] = product of everything from index i onward.
    const suf = new Array(n + 1).fill(1n);
    for (let i = n - 1; i >= 0; i--) {
        suf[i] = suf[i + 1] * BigInt(nums[i]);
    }
    const answer = [];
    // pre[i] x suf[i+1] spans everything except nums[i] itself; a lone
    // zero zeroes every cell but its own, automatically.
    for (let i = 0; i < n; i++) {
        answer.push(pre[i] * suf[i + 1]);
    }
    return answer;
};

File: problems-adapt/0001-0100/0098_product-of-the-rest/solution_prefix_arrays.go
// The judge's expected values are exact big integers, so the products are
// computed with math/big-style arithmetic (implemented directly on digit
// slices). openojRawInt serializes as a bare JSON number token via its
// MarshalJSON method; the harness marshals the return value with encoding/json.
type openojRawInt string

func (v openojRawInt) MarshalJSON() ([]byte, error) { return []byte(v), nil }

// openojBig is a signed big integer stored base 1e9, little-endian.
// An empty digit slice represents zero.
type openojBig struct {
	neg bool
	d   []uint32
}

func openojBigMulSmall(a openojBig, v int64) openojBig {
	if v == 0 || len(a.d) == 0 {
		return openojBig{}
	}
	neg := a.neg != (v < 0)
	x := uint64(v)
	if v < 0 {
		x = uint64(-v)
	}
	d := make([]uint32, 0, len(a.d)+2)
	var carry uint64
	for _, digit := range a.d {
		cur := uint64(digit)*x + carry
		d = append(d, uint32(cur%1000000000))
		carry = cur / 1000000000
	}
	for carry > 0 {
		d = append(d, uint32(carry%1000000000))
		carry /= 1000000000
	}
	return openojBig{neg: neg, d: d}
}

func openojBigMulBig(a, b openojBig) openojBig {
	if len(a.d) == 0 || len(b.d) == 0 {
		return openojBig{}
	}
	d := make([]uint32, len(a.d)+len(b.d)+1)
	for i := 0; i < len(a.d); i++ {
		var carry uint64
		j := 0
		for j < len(b.d) || carry > 0 {
			cur := uint64(d[i+j]) + carry
			if j < len(b.d) {
				cur += uint64(a.d[i]) * uint64(b.d[j])
			}
			d[i+j] = uint32(cur % 1000000000)
			carry = cur / 1000000000
			j++
		}
	}
	for len(d) > 0 && d[len(d)-1] == 0 {
		d = d[:len(d)-1]
	}
	return openojBig{neg: a.neg != b.neg, d: d}
}

func (b openojBig) String() string {
	if len(b.d) == 0 {
		return "0"
	}
	s := ""
	if b.neg {
		s = "-"
	}
	s += fmt.Sprintf("%d", b.d[len(b.d)-1])
	for i := len(b.d) - 2; i >= 0; i-- {
		s += fmt.Sprintf("%09d", b.d[i])
	}
	return s
}

func productOfRest(nums []int) []any {
	// The product except nums[i] factors as (product of everything
	// before i) x (product of everything after i), both computable as
	// running products — no division, which zeros would break anyway.
	n := len(nums)
	pre := make([]openojBig, n+1)
	// pre[i] = product of the i elements preceding index i.
	pre[0] = openojBig{d: []uint32{1}}
	for i := 0; i < n; i++ {
		pre[i+1] = openojBigMulSmall(pre[i], int64(nums[i]))
	}
	suf := make([]openojBig, n+1)
	// suf[i] = product of everything from index i onward.
	suf[n] = openojBig{d: []uint32{1}}
	for i := n - 1; i >= 0; i-- {
		suf[i] = openojBigMulSmall(suf[i+1], int64(nums[i]))
	}
	answer := make([]any, n)
	// pre[i] x suf[i+1] spans everything except nums[i] itself; a lone
	// zero zeroes every cell but its own, automatically.
	for i := 0; i < n; i++ {
		answer[i] = openojRawInt(openojBigMulBig(pre[i], suf[i+1]).String())
	}
	return answer
}
